packages/beacon-node/src/api/impl/validator/index.ts

```diff
@@ -649,223 +600,326 @@ export function getValidatorApi(
-      const result = selectBlockProductionSource({
-        builderBlockValue: builder.value.executionPayloadValue + builder.value.consensusBlockValue,
-        engineBlockValue: engine.value.executionPayloadValue + engine.value.consensusBlockValue,
-        builderBoostFactor,
-        builderSelection,
-      });
-      const executionPayloadSource = result.source;
+          return selectBuilderBlock(result.reason, {executionPayloadBlinded: true});
+        },
+      },
+    };
 
-      metrics?.blockProductionSelectionResults.inc(result);
+    return executeMatrix(engineVsBuilderMatrix, engine, builder);
+  }
 
-      logger.info(`Selected ${executionPayloadSource} block`, {
-        reason: result.reason,
-        ...loggerContext,
-        engineDurationMs: engine.durationMs,
-        ...getBlockValueLogInfo(engine.value, ProducedBlockSource.engine),
-        builderDurationMs: builder.durationMs,
-        ...getBlockValueLogInfo(builder.value, ProducedBlockSource.builder),
-      });
+  async function produceEngineOrBuilderBlock({
+    feeRecipient,
+    builderSelection,
+    strictFeeRecipientCheck,
+    slot,
+    randaoReveal,
+    graffiti,
+    builderBoostFactor,
+    forceBlockType,
+  }: routes.validator.ExtraProduceBlockOpts & {
+    slot: Slot;
+    randaoReveal: BLSSignature;
+    graffiti?: string;
+    // TODO deneb: skip randao verification
+    skipRandaoVerification?: boolean;
+    builderBoostFactor?: bigint;
+    forceBlockType?: BlockType;
+  }): Promise<ProduceFullOrBlindedBlockOrContentsRes> {
+    notWhileSyncing();
+    await waitForSlot(slot); // Must never request for a future slot > currentSlot
 
-      if (executionPayloadSource === ProducedBlockSource.engine) {
-        return {
-          ...engine.value,
-          executionPayloadBlinded: false,
-          executionPayloadSource,
-        };
+    const parentBlockRoot = fromHex(chain.getProposerHead(slot).blockRoot);
+    notOnOutOfRangeData(parentBlockRoot);
+    const ctx = normalizeBlockProductionOptions({slot, builderBoostFactor, builderSelection, forceBlockType});
+
+    const blockAttributes: BlockAttributes = {
+      feeRecipient,
+      randaoReveal,
+      graffiti: toGraffitiBuffer(
+        graffiti ?? getDefaultGraffiti(getLodestarClientVersion(opts), chain.executionEngine.clientVersion, opts)
+      ),
+      parentBlockRoot,
+      slot,
+    };
+
+    const loggerContext = {
+      slot,
+      fork: ctx.fork,
+      builderSelection,
+      strictFeeRecipientCheck,
+      // winston logger doesn't like bigint
+      builderBoostFactor: `${builderBoostFactor}`,
+    };
+
+    // Starting the process of block production
+    const blockProductionTimer = metrics?.blockProductionTime.startTimer();
+    const {block, executionPayloadValue, consensusBlockValue, executionPayloadSource, executionPayloadBlinded} =
+      await produceAssembledBlock(blockAttributes, ctx, loggerContext);
+    blockProductionTimer?.({source: executionPayloadSource});
```

Comment:
this metric becomes inaccurate because `produceAssembledBlock` is for both flow, not just for the resulting flow